src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -133,54 +146,96 @@ private static object InvokeSynchronously(string assemblyName, string methodIden
 
             var (methodInfo, parameterTypes) = GetCachedMethodInfo(assemblyName, methodIdentifier);
 
-            // There's no direct way to say we want to deserialize as an array with heterogenous
-            // entry types (e.g., [string, int, bool]), so we need to deserialize in two phases.
-            // First we deserialize as object[], for which SimpleJson will supply JsonObject
-            // instances for nonprimitive values.
-            var suppliedArgs = (object[])null;
-            var suppliedArgsLength = 0;
-            if (argsJson != null)
+            var suppliedArgs = ParseArguments(methodIdentifier, argsJson, parameterTypes);
+
+            try
             {
-                suppliedArgs = Json.Deserialize<SimpleJson.JsonArray>(argsJson).ToArray<object>();
-                suppliedArgsLength = suppliedArgs.Length;
+                return methodInfo.Invoke(targetInstance, suppliedArgs);
             }
-            if (suppliedArgsLength != parameterTypes.Length)
+            catch (TargetInvocationException tie) // Avoid using exception filters for AOT runtime support
+            {
+                if (tie.InnerException != null)
+                {
+                    ExceptionDispatchInfo.Capture(tie.InnerException).Throw();
+                    throw null; // unreached
+                }
+
+                throw;
+            }
+        }
+
+        private static object[] ParseArguments(string methodIdentifier, string argsJson, Type[] parameterTypes)
+        {
+            if (parameterTypes.Length == 0)
             {
-                throw new ArgumentException($"In call to '{methodIdentifier}', expected {parameterTypes.Length} parameters but received {suppliedArgsLength}.");
+                return Array.Empty<object>();
             }
 
-            // Second, convert each supplied value to the type expected by the method
-            var runtime = (JSRuntimeBase)JSRuntime.Current;
-            var serializerStrategy = runtime.ArgSerializerStrategy;
-            for (var i = 0; i < suppliedArgsLength; i++)
+            // There's no direct way to say we want to deserialize as an array with heterogenous
+            // entry types (e.g., [string, int, bool]), so we need to deserialize in two phases.
+            var jsonDocument = JsonDocument.Parse(argsJson);
+            var shouldDisposeJsonDocument = true;
+            try
             {
-                if (parameterTypes[i] == typeof(JSAsyncCallResult))
+                var suppliedArgsLength = jsonDocument.RootElement.GetArrayLength();
```

Comment:
worth hardening this for the case that it's not an array? I thinking about this is terms of security since this is expose via a websocket. It's good to call out and have a clear code-path for possible invalid data cases.